### DISCUSSION TITLE
Handling categoricals - best practices?

### DISCUSSION THREAD

**Discussion by K-Kielak** [meta-pytorch/botorch#1450] [General]
BoTorch uses the CategoricalKernel and MixedSingleTaskGP model that handle categoricals. This method seems to be very accurate but this accuracy seems to come at a cost in computational efficiency. As far as I understood it, the way it works is running a separate optimization process for every potential combination of categoricals what is not viable when there's a lot of them or their cardinality is high.

I wonder why this approach was used in BoTorch and I'm also trying to understand tradeoffs between other existing approaches. Namely tradeoffs between:
 1. Simple one-hot encoding
 2. One-hot encoding but setting equality constraint of 'sum of all one-hot vars for a categorical variable = 1'
 3. BoTorch way (Hamming distance for categoricals, perform a separate optimization on continuous vars for all categorical combinations)
 4. BO with Random Forest (e.g., [SMAC](https://ml.informatik.uni-freiburg.de/wp-content/uploads/papers/11-LION5-SMAC.pdf))
 5. Embedding categories using maximum likelihood estimation (as in [this paper](https://ml.informatik.uni-freiburg.de/wp-content/uploads/papers/11-LION5-SMAC.pdf))
 6. Anything else?
 
Does anyone have a good intuition on that? Are there any sources worth checking out?

**Comment by sdaulton** [comment]
> I wonder why this approach was used in BoTorch

Likely because it will work very well when it is feasible to enumerate discrete options and optimize the continuous parameters for each. It's worth checking out my new NeurIPS paper (earlier workshop version here: https://realworldml.github.io/files/cr/paper22.pdf), with many co-authors that contribute to BoTorch, which proposes a new approach for optimizing acquisition functions (AF) over discrete and mixed spaces and reviews many alternative AF optimization approaches (and different choices of surrogate models in the related work section). (The camera-ready will be out soon) Although the code is external to BoTorch at the moment (https://github.com/facebookresearch/bo_pr), I plan to put up a pull request to move key components upstream into BoTorch.

**Reply by K-Kielak** [reply]
Thanks @sdaulton, it's a great paper, I really enjoyed reading it! As promised, related work section also noted some great papers I haven't seen before.

I have a few questions after reading it though, I think the most important is just a confirmation on computational complexity. As far as I understand it, basic `optim_acqf` method from BoTorch has roughly a complexity of `O(botorch_classic)` = `O(q * num_restarts * lfbgs_steps * inputs_to_acqfval_complexity)`.

If we use `optim_acqf_mixed` it would be `O(botorch_mixed)` = `O(q * num_discrete_configs * num_restarts * lbfgs_steps * inputs_to_acqfval_complexity)` = `O(num_discrete_configs * botorch_classic)`.

Now if we use probabilistic reparametrization without MC, it would be `O(po)` = `O(q * num_restarts * lbfgs_steps * num_discrete_configs * inputs_to_acqfval_complexity)` = `O(num_discrete_configs * botorch_classic)` = `O(botorch_mixed)`, right?

So now the practical value the paper provides is when `num_discrete_configs` becomes large. At that point we can use MC approximation and have complexity of `O(mc_samples * botorch_classic)`. So instead of let's say having `O(35880*botorch_classic)` for `num_discrete_configs=35880` we can do just `O(128*botorch_classic)` without sacrificing BOs speed/convergence. Did I get the essence of it correctly, or did I miss something important?

**Reply by sdaulton** [reply]
Yeah that is the essence of it, assuming that lbfgs_steps is roughly constant. In our updated NeurIPS version (https://github.com/facebookresearch/bo_pr/blob/main/BO_Probabilistic_Reparameterization.pdf), we find that using stochastic optimization (Adam) for 200 time steps works quite well. 

In Figures 4 and 5 in Appendix E, we run an empirical evaluation akin to your theoretical complexity that compares enumeration (e.g. optim_acqf_mixed), analytic PR, and PR with 128 MC samples. The optimization performance (regret) is pretty comparable for all methods. The wall time for enumeration is by far the highest, analytic PR is fast since it uses LBFGSB and converges quickly, and MC PR is faster than enumeration, but runs Adam for 200 time steps and has roughly constant wall time over BO iterations.
![enumerate](https://user-images.githubusercontent.com/8037929/196558989-6f1bd491-1f55-414e-ba9d-b2aad1fd7e45.png)

**Comment by saitcakmak** [comment]
At a high level, the lack of certain approaches is likely due to limited bandwidth available on our side for implementing and comparing these approaches, and the approaches already implemented working well for our use cases. 

For one-hot encoding based approaches, [Ax](ax.dev) supports one-hot encoding of categoricals. 

> Embedding categories using maximum likelihood estimation (as in [this paper](https://ml.informatik.uni-freiburg.de/wp-content/uploads/papers/11-LION5-SMAC.pdf))

See models defined in https://github.com/pytorch/botorch/blob/main/botorch/models/contextual.py and https://github.com/pytorch/botorch/blob/main/botorch/models/contextual_multioutput.py for embedding based approaches. I believe @qingfeng10 has worked with these, but I don't know if we're actively using them anywhere.

**Reply by K-Kielak** [reply]
Thank you @saitcakmak for the reply, I really appreciate it!

Of course, I understand that with unlimited bandwidth there would be more options I was just curious why specifically this approach was implemented and not one of the others. But I guess it makes sense, the approach implemented in BoTorch is as accurate as it gets while not being computationally efficient, but compute is abundant at Meta so computational cost is not a big problem :sweat_smile: Tbf, I think it may work for us as well in the end but I'm still exploring other options.

Regarding embeddings, I'm curious about general effectiveness of these methods. I've seen a couple of different approaches but I lack an intuition on their unbiased strenghts/weaknesses and when they are appropriate. Do you know of any papers/sources worth checking out in that regard?

**Reply by saitcakmak** [reply]
@K-Kielak I'm not too familiar with the literature on this, but I'd check out this paper: https://proceedings.neurips.cc/paper/2020/hash/faff959d885ec0ecf70741a846c34d1d-Abstract.html
